### PR TITLE
Support Arbitrary JSON values in JSON Reader (#4905)

### LINKED ISSUE #4905: Support for reading JSON Array to Arrow
**Which part is this question about**

Reading from JSON to Arrow ([arrow_json](https://docs.rs/arrow-json/47.0.0/arrow_json/index.html)::[reader](https://docs.rs/arrow-json/47.0.0/arrow_json/reader/index.html#) module)

**Describe your question**

[arrow_json::reader](https://docs.rs/arrow-json/47.0.0/arrow_json/reader/index.html#) mentions
> This JSON reader allows **JSON line-delimited files** to be read into the Arrow memory model...

I'm curious as to why does the reader not support Array JSON type (`[{...}, {...}, {...}, ...]`)? I found that `arrow2` is able to [do so](https://jorgecarleitao.github.io/arrow2/io/json_read.html) but not `arrow-rs`.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4905

# Rationale for this change
 

We should support decoding JSON payloads that don't have an object as the root.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
This does represent a fairly minor performance regression, but I'm not too concerned

```
small_bench_primitive   time:   [7.0711 µs 7.0744 µs 7.0778 µs]
                        change: [+2.9718% +3.2020% +3.6122%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) low mild
  2 (2.00%) high mild
  3 (3.00%) high severe

large_bench_primitive   time:   [2.4153 ms 2.4161 ms 2.4169 ms]
                        change: [+5.7843% +5.8318% +5.8805%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) high mild
  9 (9.00%) high severe

small_bench_list        time:   [13.203 µs 13.211 µs 13.220 µs]
                        change: [+1.2264% +1.3806% +1.5824%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe
```

File: arrow-json/src/reader/mod.rs [hunk 1]
```diff
@@ -17,9 +17,13 @@
...
-//! This JSON reader allows JSON line-delimited files to be read into the Arrow memory
-//! model. Records are loaded in batches and are then converted from row-based data to
-//! columnar data.
+//! This JSON reader allows JSON records to be read into the Arrow memory
+//! model. Records are loaded in batches and are then converted from the record-oriented
+//! representation to the columnar arrow data model.
+//!
+//! The reader is agnostic to whitespace, including `\n` and `\r`, and will ignore such characters.
+//! This allows parsing sequences of one or more arbitrarily formatted JSON values, including
+//! but not limited to newline-delimited JSON.
```
Comment (alamb) [inline arrow-json/src/reader/mod.rs:26]: A minor suggestion to improve the wording here

```suggestion
//! The reader ignores whitespace between JSON values, including `\n` and `\r`.
//! This allows parsing sequences of one or more arbitrarily formatted JSON values, including
//! but not limited to newline-delimited JSON.
```

File: arrow-json/src/reader/mod.rs [hunk 2]
```diff
@@ -189,10 +197,51 @@ impl ReaderBuilder {
             batch_size: 1024,
             coerce_primitive: false,
             strict_mode: false,
+            is_field: false,
             schema,
         }
     }
 
+    /// Create a new [`ReaderBuilder`] with the provided [`FieldRef`]
```
Comment (alamb) [inline arrow-json/src/reader/mod.rs:205]:
```suggestion
    /// Create a new [`ReaderBuilder`] that will parse JSON values with a root schema of [`FieldRef`].
```

Perhaps we can add a note in `new()` that says it does require the root to be an object like `{..}`.

I wonder if `new_from_field` might be a more descriptive name (as this isn't making a new Field, it is making a new reader that reads data with the type on the field

File: arrow-json/src/reader/mod.rs [hunk 3]
```diff
@@ -233,16 +282,22 @@ impl ReaderBuilder {
...
     /// Create a [`Decoder`]
     pub fn build_decoder(self) -> Result<Decoder, ArrowError> {
-        let decoder = make_decoder(
-            DataType::Struct(self.schema.fields.clone()),
-            self.coerce_primitive,
-            self.strict_mode,
-            false,
-        )?;
+        let (data_type, nullable) = match self.is_field {
+            false => (DataType::Struct(self.schema.fields.clone()), false),
```
Comment (alamb) [inline arrow-json/src/reader/mod.rs:286]: why not allow null root fields with structs?

File: arrow-json/src/reader/tape.rs [hunk 4]
```diff
@@ -297,7 +297,7 @@ macro_rules! next {
 pub struct TapeDecoder {
     elements: Vec<TapeElement>,
 
-    num_rows: usize,
+    cur_row: usize,
```
Comment (alamb) [inline arrow-json/src/reader/tape.rs:300]:
```suggestion
    /// logical row being decoded 
    cur_row: usize,
```

**Review by alamb — APPROVED** [review]
Looks really solid to me -- thank you @tustvold 

I had some documentation suggestions, but nothing I think that is required before merging

File: arrow-json/src/reader/mod.rs (hunk 3 above)
Comment (tustvold) [reply arrow-json/src/reader/mod.rs:286]: Because RecordBatch can't support nulls at the root level

**Pull request merged**